### PR TITLE
Removed root browsing context from constellation

### LINKED ISSUE #13994: [multi window] Constellation should be able to handle multiple root frame
At the moment, a constellation has only one root frame.

When Servo will support multiple windows, we would need multiple root frames (and multiple compositors, see #13993). We might also want to have multiple root frames for one compositor (non-mozbrowser based embedder will have one root frame per tab).

### PR DESCRIPTION
**Pull request by asajeffrey**
Removed the special root browsing context from the constellation.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13994 
- [X] These changes do not require tests because this isn't visible from user code

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17077)

### CODE REVIEW

**Comment by paulrouget** [comment]
I'm looking at this. It will take me a day or two. I'd like to see how close we are to support tabs with this.

**Comment by asajeffrey** [comment]
Thanks! Let me know if there's anything else needed for a minimal one-window-but-multiple-tabs browser.

**Comment by paulrouget** [comment]
So now, if I'm not mistaken, the compositor has a pipeline tree while the constellation has a pipeline forest. That means the compositor will get plenty of messages from or for pipeline it doesn't know about. I think it brings us back to #15934 (Create a dedicated channel for constellation -> embedder communication).

Can we land support for multiple browsing contexts without fixing #15934 first?

**Comment by paulrouget** [comment]
@asajeffrey I managed to get tabs working. Thank you! The only issue is what I mentioned in my previous comment, but I will take care of that later.

r=me, but you need a proper review though (I'm not a reviewer).

Once this has landed, this is what I will work on:
1. by-pass the compositor for non-compositor message. See my plan here: https://github.com/servo/servo/issues/15934#issuecomment-306133089
2. expose TopLevelBrowsingContext to embedder
3. implement API to create and switch browsing context
4. add a browsing context parameter to the WindowEvents (when it makes sense)
5. update the visibility API status based on the active compositor pipeline tree

I got most of this working here: https://github.com/paulrouget/servo/tree/tabs (just prototyping, code is horrible).

**Comment by paulrouget** [comment]
@asajeffrey don't you think we should clear `IOCompositor::pipeline_details` when the frame tree is set?

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -1188,10 +1177,16 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
 
         self.mem_profiler_chan.send(mem::ProfilerMsg::Exit);
 
-        // TODO: exit before the root browsing context is initialized?
-        debug!("Removing root browsing context.");
-        let root_browsing_context_id = BrowsingContextId::from(self.root_browsing_context_id);
-        self.close_browsing_context(root_browsing_context_id, ExitPipelineMode::Normal);
+        // Close the top-level browsing contexts
+        let top_level_browsing_context_ids: Vec<TopLevelBrowsingContextId> = self.browsing_contexts.values()
+            .filter(|browsing_context| browsing_context.is_top_level())
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: nit: `.filter(BrowsingContext::is_top_level)`

File: components/constellation/constellation.rs [hunk 2]
```diff
@@ -1188,10 +1177,16 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
+            .map(|browsing_context| browsing_context.top_level_id)
+            .collect();
+        for top_level_browsing_context_id in top_level_browsing_context_ids {
+            debug!("Removing top-level browsing context {}.", top_level_browsing_context_id);
+            let browsing_context_id = BrowsingContextId::from(top_level_browsing_context_id);
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: If we know the browsing contexts are top level, can't we just get their `id` rather than the `top_level_id` to avoid this conversion?

File: components/constellation/constellation.rs [hunk 3]
```diff
@@ -1495,6 +1490,7 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         };
 
         // Create the new pipeline, attached to the parent and push to pending changes
+        self.handle_load_start_msg(load_info.info.new_pipeline_id);
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: If we need to be calling `self.handle_load_start_msg` before we push a pending change, it might make sense to add a new method for adding pending changes that way we don't forget to call `self.handle_load_start_msg`.

This could be saved for another PR since it isn't directly related to these changes. Possible E-Easy.

File: components/constellation/constellation.rs [hunk 4]
```diff
@@ -1680,14 +1682,12 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
                 if let Err(e) = result {
                     self.handle_send_error(parent_pipeline_id, e);
                 }
-                Some(source_id)
+                None
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: I'm not very familiar with webdriver, but its the only user of the pipeline id returned from this method. Does this change break anything there?

File: components/constellation/constellation.rs [hunk 5]
```diff
@@ -1657,14 +1653,20 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         let (browsing_context_id, parent_info) = match self.pipelines.get(&source_id) {
             Some(pipeline) => (pipeline.browsing_context_id, pipeline.parent_info),
             None => {
-                warn!("Pipeline {:?} loaded after closure.", source_id);
+                warn!("Pipeline {} loaded after closure.", source_id);
+                return None;
+            }
+        };
+        let (top_level_id, window_size, timestamp) = match self.browsing_contexts.get(&browsing_context_id) {
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: These values are only used in the `None` branch of the match, can we move them there?

File: components/constellation/constellation.rs [hunk 6]
```diff
@@ -2290,7 +2280,9 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             self.focus_pipeline_id = Some(change.new_pipeline_id);
         }
 
-        let (evicted_id, new_context, navigated, location_changed) = if let Some(instant) = change.replace_instant {
+        let url = change.load_data.url.clone();
+
+        let (evicted_id, new_context, navigated) = if let Some(instant) = change.replace_instant {
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: Yay one less item in this tuple!

**Review by cbrewster — CHANGES_REQUESTED** [review]
Looks good other than a couple nits/questions!

File: components/constellation/constellation.rs (hunk 1 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: We wish. It's got the wrong type (`&BrowsingContext` vs `BrowsingContext`).

File: components/constellation/constellation.rs (hunk 2 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Fixed.

File: components/constellation/constellation.rs (hunk 3 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Yes, there's quite a bit of boilerplate we could DRY. This should be a separate PR though. Not sure anything in the constellation is E-Easy :)

File: components/constellation/constellation.rs (hunk 5 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Fixed.

File: components/constellation/constellation.rs (hunk 4 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Er, I don't think so. webdriver isn't very tested though :/

File: components/constellation/constellation.rs (hunk 6 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Yay!

**Review by asajeffrey — COMMENTED** [review]
Pushed changes.

File: components/constellation/constellation.rs (hunk 1 above)
Comment (cbrewster) [inline components/constellation/constellation.rs]: Tis a shame

**Review by cbrewster — APPROVED** [review]
LGTM. squash and r=me

**Comment by cbrewster** [comment]
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/mozbrowser/redirect.html:
  │ FAIL [expected PASS] mozbrowserlocationchange with final url after redirect
  │   → assert_equals: expected "about:blank" but got "http://web-platform.test:8000/_mozilla/mozilla/mozbrowser/redirect_init.html?pipe=status(302)|header(Location,redirect_final.html)"
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/mozbrowser/redirect.html:16:21
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20
```

**Comment by asajeffrey** [comment]
Oh joy, I remember this, the question is whether `mozbrowserlocationchange` should be fired with the initial `about:blank` or not.

**Comment by asajeffrey** [comment]
IRC conversation with @cbrewster: http://logs.glob.uno/?c=mozilla%23servo&s=7+Jun+2017&e=7+Jun+2017#c691847

@bors-servo r=cbrewster

**Comment by cbrewster** [comment]
```
Checking files for tidiness...

./components/constellation/constellation.rs:2891: Line is longer than 120 characters
```

**Comment by asajeffrey** [comment]
Oh for pity's sake...
